src/Trains.NET.Rendering/LayerRenderer/TerrainTypeRenderer.cs

```diff
@@ -0,0 +1,196 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Trains.NET.Engine;
+using Trains.NET.Engine.Terrain;
+using Trains.NET.Rendering.Terrain;
+
+namespace Trains.NET.Rendering
+{
+    [Order(0)]
+    internal class TerrainTypeRenderer : ILayerRenderer
+    {
+        private readonly ITerrainMap _terrainMap;
+        private readonly IPixelMapper _pixelMapper;
+        private readonly ITrackParameters _trackParameters;
+        private readonly IGradientRenderer _gradientRenderer;
+
+        public TerrainTypeRenderer(ITerrainMap terrainMap, IPixelMapper pixelMapper, ITrackParameters trackParameters, IGradientRenderer gradientRenderer)
+        {
+            _terrainMap = terrainMap;
+            _pixelMapper = pixelMapper;
+            _trackParameters = trackParameters;
+            _gradientRenderer = gradientRenderer;
+        }
+
+        public bool Enabled { get; set; } = true;
+        public string Name => "TerrainType";
+
+        public void Render(ICanvas canvas, int width, int height)
+        {
+            DrawAllTerrainCells(canvas);
+        }
+
+        private void DrawAllTerrainCells(ICanvas canvas)
+        {
+            foreach (var terrain in _terrainMap)
+            {
+                DrawTerrainCell(canvas, terrain);
+            }
+        }
+
+        private void DrawTerrainCell(ICanvas canvas, TerrainCell terrain)
+        {
+            var orthogonalGradients = new Dictionary<TerrainAdjacency, TerrainCellGradient>();
+            // Find all edge gradients
+            foreach(var adjacency in AdjacencyTypes.OrthogonalAdjacents)
+            {
+                var gradient = GenerateEdgeGradient(terrain, adjacency);
+                orthogonalGradients[adjacency] = gradient;
+            }
+
+            var cornerGradients = new Dictionary<TerrainAdjacency, TerrainCellGradient>();
+            // Use edge transitions to generate corner gradients
+            foreach (var adjacency in AdjacencyTypes.CornerAdjacents)
+            {
+                var gradient = GenerateGradientForCornerCell(terrain, adjacency, orthogonalGradients);
+                if (gradient.GradientType == GradientType.None) continue;
+                cornerGradients[adjacency] = gradient;
+            }
+
+            RenderBackground(terrain, canvas);
+
+            // Render edge gradients, highest weight last
+            foreach (var gradient in orthogonalGradients.Values.Where(g => g.GradientType != GradientType.None).OrderBy(g => GetWeightForGradientType(g.GradientType))) {
+                _gradientRenderer.Render(canvas, gradient);
```

Comment:
Maybe EdgeGradientRenderer and CornerGradientRenderer as distinct services may make more sense